### PR TITLE
Add revoke button to API keys page

### PR DESCRIPTION
**Pull request by mert019**
* Revoke button added for api keys.
* Existing revocation mechanism used.

Addresses https://github.com/NuGet/NuGetGallery/issues/9426

edit by @joelverhagen -- added screenshot
image
image

### CODE REVIEW

**Comment by mert019** [comment]
> @dotnet-policy-service agree

**Comment by mert019** [comment]
@dotnet-policy-bot agree

**Comment by joelverhagen** [comment]
This looks great @mert019. Thanks for taking the time to work on this. We'll review it and run CI now!

**Comment by joelverhagen** [comment]
@mert019, it seems like there is some issue with the CLA bot. Could you try re-opening the PR per this suggestion? https://github.com/microsoft/ContributorLicenseAgreement/issues/123
Sorry about the inconvenience.

**Comment by mert019** [comment]
@dotnet-policy-service agree

**Comment by joelverhagen** [comment]
Hmmm ... weird.

In https://github.com/NuGet/NuGetGallery/pull/9509#issuecomment-1555951430 -- this was a quote, so it maybe didn't work because of that? Guessing here.

In https://github.com/NuGet/NuGetGallery/pull/9509#issuecomment-1555958001 -- this is @'ing the wrong username (@dotnet-policy-bot instead of @dotnet-policy-service).

**Comment by joelverhagen** [comment]
That did it. Thanks!

File: src/NuGetGallery.Services/Authentication/AuthenticationService.cs [hunk 1]
```diff
@@ -178,7 +178,7 @@ public Credential GetApiKeyCredential(string apiKey)
             return FindMatchingApiKey(credential);
         }
 
-        public async Task RevokeApiKeyCredential(Credential apiKeyCredential, CredentialRevocationSource revocationSourceKey, bool commitChanges = true)
+        public virtual async Task RevokeApiKeyCredential(Credential apiKeyCredential, CredentialRevocationSource revocationSourceKey, bool commitChanges = true)
```
Comment (agr) [inline src/NuGetGallery.Services/Authentication/AuthenticationService.cs:181]: Why was it marked as `virtual`?

Comment (mert019) [reply src/NuGetGallery.Services/Authentication/AuthenticationService.cs:181]: To mock this method in tests, I had to mark it as virtual.

**Comment by erdembayar** [comment]
LGTM. But I can't sign off, I didn't do js reviews for while.

File: src/NuGetGallery/Controllers/UsersController.cs [hunk 2]
```diff
@@ -897,6 +897,33 @@ public virtual async Task<ActionResult> RemoveCredential(string credentialType,
...
+
+            if (cred == null)
+            {
+                Response.StatusCode = (int)HttpStatusCode.NotFound;
+                return Json(Strings.CredentialNotFound);
+            }
+
+            if (CredentialTypes.IsApiKey(credentialType))
+            {
+                await AuthenticationService.RevokeApiKeyCredential(cred, CredentialRevocationSource.User, true);
```
Comment (agr) [inline src/NuGetGallery/Controllers/UsersController.cs:919]: An email notification has to be sent here similar to how `RemoveApiKeyCredential` function does it.

File: tests/NuGetGallery.Facts/Controllers/UsersControllerFacts.cs [hunk 3]
```diff
@@ -2221,6 +2221,78 @@ public static IEnumerable<object[]> GivenValidRequest_ItGeneratesNewCredAndRemov
...
+                // Assert
+                Assert.IsType<JsonResult>(result);
+                var credentialViewModel = Assert.IsType<ApiKeyViewModel>(result.Data);
+                Assert.True(credentialViewModel.HasExpired);
+
+                authenticationService.Verify(x => x.RevokeApiKeyCredential(It.IsAny<Credential>(), It.IsAny<CredentialRevocationSource>(), It.IsAny<bool>()), Times.Once);
+                authenticationService.Verify(x => x.DescribeCredential(It.IsAny<Credential>()), Times.Once);
+            }
+
+            private CredentialViewModel GetRevokedApiKeyCredentialViewModel(string apiKeyType)
```
Comment (agr) [inline tests/NuGetGallery.Facts/Controllers/UsersControllerFacts.cs:2284]: Can it be made `static`?

File: tests/NuGetGallery.Facts/Controllers/UsersControllerFacts.cs [hunk 4]
```diff
@@ -2221,6 +2221,78 @@ public static IEnumerable<object[]> GivenValidRequest_ItGeneratesNewCredAndRemov
...
+
+                // Act
+                var result = await controller.RevokeCredential(
+                    credentialType: cred.Type,
+                    credentialKey: CredentialKey);
+
+                // Assert
+                Assert.Equal((int)HttpStatusCode.NotFound, controller.Response.StatusCode);
+                Assert.Equal(Strings.CredentialNotFound, (string)result.Data);
+                Assert.Equal(1, user.Credentials.Count);
```
Comment (agr) [inline tests/NuGetGallery.Facts/Controllers/UsersControllerFacts.cs:2247]: nit: `Assert.Single(user.Credentials)`

**Comment by mert019** [comment]
Currently I am busy with my final exams. I will commit the changes when I have free time.

**Comment by joelverhagen** [comment]
Good luck on your finals! 💯

**Comment by mert019** [comment]
Hi @agr
I pushed new changes. Can you please review it?

**Comment by agr** [comment]
Apologies for the delay. I am looking.

File: src/NuGetGallery/Strings.resx [hunk 5]
```diff
@@ -1254,4 +1254,10 @@ The {1} Team</value>
   <data name="ForgotPassword_Disabled_Error" xml:space="preserve">
     <value>Forgot password is disabled.</value>
   </data>
+  <data name="Emails_ApiKeyRevoked_Body" xml:space="preserve">
+    <value>Your API Key '{0}' revoked. If you did not request this change, please reply to this email to contact support.</value>
```
Comment (agr) [inline src/NuGetGallery/Strings.resx:1258]:
```suggestion
    <value>Your API Key '{0}' has been revoked. If you did not request this change, please reply to this email to contact support.</value>

```

Comment (agr) [reply src/NuGetGallery/Strings.resx:1258]: That would also require fixing a test below, I believe.

**Review by agr — DISMISSED** [review]
Looks good to me. Thank you for the contribution!
I added one more comment though, please take a look.
I'll wait for other reviews for couple of days and then will look into merging it.

File: src/NuGetGallery/Strings.resx (hunk 5 above)
Comment (mert019) [reply src/NuGetGallery/Strings.resx]: Revised

**Review by joelverhagen — APPROVED** [review]

**Review by agr — APPROVED** [review]

**Pull request merged**